dlib/cuda/tensor_abstract.h

```diff
@@ -718,6 +718,36 @@ namespace dlib
         provides serialization support for alias_tensor.  
     !*/
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename label_type>
+    struct weighted_label
```

Comment:
Yeah, this is a weird place for it.  Put it in dnn/utilities.h instead.  It probably makes the most sense to leave the documentation for it in loss.h though since that's right next to the docs that depend on it, so it will be easily located by readers.  We are only talking about moving this because of include order stuff after all, not because it logically makes sense to take it out of loss.h

Reply:
Agreed. Maybe I'm missing something, but if I just put it in `dnn/utilities.h` and `#include "../dnn/utilites.h" in the `cuda/cuda_dlib.h` I get a compilation error:
```
╭───────────────────────────────────────────────────────────────────────────────────────╮
│ • In included file: main file cannot be included recursively when building a preamble │
│                                                                                       │
│ /home/adria/Projects/dlib/dlib/cuda/../dnn/../cuda/tensor_tools.h:12:10:              │
│ note: error occurred here                                                             │
╰───────────────────────────────────────────────────────────────────────────────────────╯
```
That is why I put it outside of the `dnn` directory... maybe add a new file in the `cuda` directory? It's really weird...